File: RentalApp_FASTAPI/shared/services/period_service.py
"""
Сервис для работы с временными периодами.
Предоставляет функциональность для вычисления дат начала и окончания периодов.
"""

from datetime import datetime, date, timedelta
from typing import Optional, Tuple, Dict, Any
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class PeriodType(str, Enum):
    """Типы временных периодов."""
    WEEK = "week"
    MONTH = "month"
    QUARTER = "quarter"
    YEAR = "year"


class PeriodService:
    """Сервис для работы с временными периодами."""
    
    def get_period_dates(self, period_type: str, period_offset: int = 0) -> Tuple[date, date]:
        """
        Вычисляет даты начала и окончания для указанного периода.
        
        Args:
            period_type: Тип периода (week, month, quarter, year)
            period_offset: Смещение периода (0 = текущий, -1 = предыдущий, 1 = следующий)
            
        Returns:
            Кортеж (дата начала, дата окончания)
            
        Raises:
            ValueError: Если указан неподдерживаемый тип периода
        """
        try:
            period_type_enum = PeriodType(period_type)
        except ValueError:
            raise ValueError(f"Неподдерживаемый тип периода: {period_type}")
        
        today = date.today()
        
        if period_type_enum == PeriodType.WEEK:
            return self._get_week_dates(today, period_offset)
        elif period_type_enum == PeriodType.MONTH:
            return self._get_month_dates(today, period_offset)
        elif period_type_enum == PeriodType.QUARTER:
            return self._get_quarter_dates(today, period_offset)
        elif period_type_enum == PeriodType.YEAR:
            return self._get_year_dates(today, period_offset)
    
    def _get_week_dates(self, base_date: date, offset: int) -> Tuple[date, date]:
        """Вычисляет даты для недельного периода."""
        # Находим понедельник текущей недели
        days_since_monday = base_date.weekday()
        monday = base_date - timedelta(days=days_since_monday)
        
        # Применяем смещение
        target_monday = monday + timedelta(weeks=offset)
        target_sunday = target_monday + timedelta(days=6)
        
        return target_monday, target_sunday
    
    def _get_month_dates(self, base_date: date, offset: int) -> Tuple[date, date]:
        """Вычисляет даты для месячного периода."""
        # Находим первый день текущего месяца
        first_day = base_date.replace(day=1)
        
        # Применяем смещение
        if offset >= 0:
            for _ in range(offset):
                if first_day.month == 12:
                    first_day = first_day.replace(year=first_day.year + 1, month=1)
                else:
                    first_day = first_day.replace(month=first_day.month + 1)
        else:
            for _ in range(abs(offset)):
                if first_day.month == 1:
                    first_day = first_day.replace(year=first_day.year - 1, month=12)
                else:
                    first_day = first_day.replace(month=first_day.month - 1)
        
        # Находим последний день месяца
        if first_day.month == 12:
            next_month = first_day.replace(year=first_day.year + 1, month=1)
        else:
            next_month = first_day.replace(month=first_day.month + 1)
        
        last_day = next_month - timedelta(days=1)
        
        return first_day, last_day
    
    def _get_quarter_dates(self, base_date: date, offset: int) -> Tuple[date, date]:
        """Вычисляет даты для квартального периода."""
        # Определяем текущий квартал
        current_quarter = (base_date.month - 1) // 3 + 1
        quarter_start_month = (current_quarter - 1) * 3 + 1
        
        # Находим первый день текущего квартала
        first_day = base_date.replace(month=quarter_start_month, day=1)
        
        # Применяем смещение
        if offset >= 0:
            for _ in range(offset):
                if first_day.month <= 9:
                    first_day = first_day.replace(month=first_day.month + 3)
                else:
                    first_day = first_day.replace(year=first_day.year + 1, month=1)
        else:
            for _ in range(abs(offset)):
                if first_day.month > 3:
                    first_day = first_day.replace(month=first_day.month - 3)
                else:
                    first_day = first_day.replace(year=first_day.year - 1, month=10)
        
        # Находим последний день квартала
        if first_day.month <= 9:
            next_quarter_start = first_day.replace(month=first_day.month + 3)
        else:
            next_quarter_start = first_day.replace(year=first_day.year + 1, month=1)
        
        last_day = next_quarter_start - timedelta(days=1)
        
        return first_day, last_day
    
    def _get_year_dates(self, base_date: date, offset: int) -> Tuple[date, date]:
        """Вычисляет даты для годового периода."""
        # Находим первый день текущего года
        first_day = base_date.replace(month=1, day=1)
        
        # Применяем смещение
        target_year = first_day.year + offset
        first_day = first_day.replace(year=target_year)
        
        # Последний день года
        last_day = first_day.replace(month=12, day=31)
        
        return first_day, last_day
    
    def get_period_label(self, period_type: str, period_offset: int = 0) -> str:
        """
        Генерирует читаемую метку для периода.
        
        Args:
            period_type: Тип периода
            period_offset: Смещение периода
            
        Returns:
            Строка с описанием периода
        """
        start_date, end_date = self.get_period_dates(period_type, period_offset)
        
        if period_type == PeriodType.WEEK:
            return f"Неделя {start_date.strftime('%d.%m')} - {end_date.strftime('%d.%m.%Y')}"
        elif period_type == PeriodType.MONTH:
            return f"{start_date.strftime('%B %Y')}"
        elif period_type == PeriodType.QUARTER:
            quarter_num = (start_date.month - 1) // 3 + 1
            return f"Q{quarter_num} {start_date.year}"
        elif period_type == PeriodType.YEAR:
            return f"{start_date.year}"
        
        return f"Период {start_date} - {end_date}"
    
    def validate_period_params(self, period_type: Optional[str], period_offset: int = 0) -> bool:
        """
        Валидирует параметры периода.
        
        Args:
            period_type: Тип периода
            period_offset: Смещение периода
            
        Returns:
            True если параметры валидны, False иначе
        """
        if period_type is None:
            return True  # Период не задан - это валидно
        
        try:
            PeriodType(period_type)
        except ValueError:
            return False
        
        # Проверяем разумные пределы смещения
        if abs(period_offset) > 100:  # Ограничиваем смещение разумными пределами
            return False
        
        return True
    
    def get_period_info(self, period_type: str, period_offset: int = 0) -> Dict[str, Any]:
        """
        Возвращает полную информацию о периоде.
        
        Args:
            period_type: Тип периода
            period_offset: Смещение периода
            
        Returns:
            Словарь с информацией о периоде
        """
        start_date, end_date = self.get_period_dates(period_type, period_offset)
        
        return {
            "type": period_type,
            "offset": period_offset,
            "start_date": start_date,
            "end_date": end_date,
            "label": self.get_period_label(period_type, period_offset),
            "days_count": (end_date - start_date).days + 1
        }

File: rental-app-main/src/core/services/PeriodService.ts
/**
 * Сервис для работы с временными периодами во frontend
 */

import type {
  PeriodType,
  PeriodOption,
  PeriodServiceResult,
  PeriodNavigation
} from "@/types/period";

export class PeriodService {
  /**
   * Вычисляет даты начала и окончания для указанного периода
   */
  static getPeriodDates(periodType: PeriodType, periodOffset: number = 0): { startDate: Date; endDate: Date } {
    const today = new Date();

    switch (periodType) {
      case "week":
        return this.getWeekDates(today, periodOffset);
      case "month":
        return this.getMonthDates(today, periodOffset);
      case "quarter":
        return this.getQuarterDates(today, periodOffset);
      case "year":
        return this.getYearDates(today, periodOffset);
      default:
        throw new Error(`Неподдерживаемый тип периода: ${periodType}`);
    }
  }

  /**
   * Вычисляет даты для недельного периода
   */
  private static getWeekDates(baseDate: Date, offset: number): { startDate: Date; endDate: Date } {
    // Находим понедельник текущей недели
    const dayOfWeek = baseDate.getDay();
    const daysSinceMonday = dayOfWeek === 0 ? 6 : dayOfWeek - 1; // Воскресенье = 0, понедельник = 1
    const monday = new Date(baseDate);
    monday.setDate(baseDate.getDate() - daysSinceMonday);

    // Применяем смещение
    const targetMonday = new Date(monday);
    targetMonday.setDate(monday.getDate() + (offset * 7));

    const targetSunday = new Date(targetMonday);
    targetSunday.setDate(targetMonday.getDate() + 6);

    return { startDate: targetMonday, endDate: targetSunday };
  }

  /**
   * Вычисляет даты для месячного периода
   */
  private static getMonthDates(baseDate: Date, offset: number): { startDate: Date; endDate: Date } {
    const firstDay = new Date(baseDate.getFullYear(), baseDate.getMonth(), 1);

    // Применяем смещение
    const targetMonth = new Date(firstDay);
    targetMonth.setMonth(firstDay.getMonth() + offset);

    const startDate = new Date(targetMonth.getFullYear(), targetMonth.getMonth(), 1);
    const endDate = new Date(targetMonth.getFullYear(), targetMonth.getMonth() + 1, 0);

    return { startDate, endDate };
  }

  /**
   * Вычисляет даты для квартального периода
   */
  private static getQuarterDates(baseDate: Date, offset: number): { startDate: Date; endDate: Date } {
    const currentQuarter = Math.floor(baseDate.getMonth() / 3);
    const quarterStartMonth = currentQuarter * 3;

    const firstDay = new Date(baseDate.getFullYear(), quarterStartMonth, 1);

    // Применяем смещение
    const targetQuarter = new Date(firstDay);
    targetQuarter.setMonth(firstDay.getMonth() + (offset * 3));

    const startDate = new Date(targetQuarter.getFullYear(), targetQuarter.getMonth(), 1);
    const endDate = new Date(targetQuarter.getFullYear(), targetQuarter.getMonth() + 3, 0);

    return { startDate, endDate };
  }

  /**
   * Вычисляет даты для годового периода
   */
  private static getYearDates(baseDate: Date, offset: number): { startDate: Date; endDate: Date } {
    const targetYear = baseDate.getFullYear() + offset;

    const startDate = new Date(targetYear, 0, 1);
    const endDate = new Date(targetYear, 11, 31);

    return { startDate, endDate };
  }

  /**
   * Генерирует читаемую метку для периода
   */
  static getPeriodLabel(periodType: PeriodType | null, periodOffset: number = 0): string {
    if (!periodType) return "Все периоды";
    if (!this.validatePeriodParams(periodType, periodOffset)) return "Ошибка периода";

    try {
      const { startDate, endDate } = this.getPeriodDates(periodType, periodOffset);

      switch (periodType) {
        case "week":
          return `Неделя ${this.formatDate(startDate, "dd.MM")} - ${this.formatDate(endDate, "dd.MM.yyyy")}`;
        case "month":
          return this.formatDate(startDate, "MMMM yyyy");
        case "quarter":
          const quarterNum = Math.floor(startDate.getMonth() / 3) + 1;
          return `${quarterNum} Квартал ${startDate.getFullYear()}`;
        case "year":
          return startDate.getFullYear().toString();
        default:
          return `${this.formatDate(startDate, "dd.MM.yyyy")} - ${this.formatDate(endDate, "dd.MM.yyyy")}`;
      }
    } catch {
      return "Ошибка периода";
    }
  }

  /**
   * Форматирует дату в указанном формате
   */
  private static formatDate(date: Date, format: string): string {
    const months = [
      "января", "февраля", "марта", "апреля", "мая", "июня",
      "июля", "августа", "сентября", "октября", "ноября", "декабря"
    ];

    const day = date.getDate().toString().padStart(2, "0");
    const month = (date.getMonth() + 1).toString().padStart(2, "0");
    const year = date.getFullYear();
    const monthName = months[date.getMonth()];

    return format
      .replace("dd", day)
      .replace("MM", month)
      .replace("yyyy", year.toString())
      .replace("MMMM", monthName);
  }

  /**
   * Возвращает полную информацию о периоде
   */
  static getPeriodInfo(periodType: PeriodType, periodOffset: number = 0): PeriodServiceResult {
    const { startDate, endDate } = this.getPeriodDates(periodType, periodOffset);
    const label = this.getPeriodLabel(periodType, periodOffset);

    return {
      startDate,
      endDate,
      label
    };
  }

  /**
   * Валидирует параметры периода
   */
  static validatePeriodParams(periodType: PeriodType | null, periodOffset: number = 0): boolean {
    if (periodType === null) {
      return true; // Период не задан - это валидно
    }

    const validTypes: PeriodType[] = ["week", "month", "quarter", "year"];
    if (!validTypes.includes(periodType)) {
      return false;
    }

    // Проверяем разумные пределы смещения
    if (Math.abs(periodOffset) > 100) {
      return false;
    }

    return true;
  }

  /**
   * Возвращает опции периодов для UI
   */
  static getPeriodOptions(): PeriodOption[] {
    return [
      {
        value: "week",
        label: "Неделя",
        description: "Показать резервы/аренды за неделю"
      },
      {
        value: "month",
        label: "Месяц",
        description: "Показать резервы/аренды за месяц"
      },
      {
        value: "quarter",
        label: "Квартал",
        description: "Показать резервы/аренды за квартал"
      },
      {
        value: "year",
        label: "Год",
        description: "Показать резервы/аренды за год"
      }
    ];
  }

  /**
   * Возвращает информацию о навигации по периодам
   */
  static getPeriodNavigation(periodType: PeriodType | null, periodOffset: number = 0): PeriodNavigation {
    if (!periodType) {
      return {
        canGoBack: false,
        canGoForward: false,
        currentOffset: 0,
        currentLabel: "Все периоды"
      };
    }

    const currentLabel = this.getPeriodLabel(periodType, periodOffset);

    return {
      canGoBack: true,
      canGoForward: true,
      currentOffset: periodOffset,
      currentLabel
    };
  }

  /**
   * Проверяет, пересекается ли период с заданными датами
   */
  static isPeriodOverlapping(
    periodType: PeriodType,
    periodOffset: number,
    startDate: Date,
    endDate: Date
  ): boolean {
    const { startDate: periodStart, endDate: periodEnd } = this.getPeriodDates(periodType, periodOffset);

    // Проверяем пересечение: startDate <= periodEnd AND endDate >= periodStart
    return startDate <= periodEnd && endDate >= periodStart;
  }
}
